nrf52-hal-common/src/spim.rs

```diff
@@ -41,77 +37,83 @@ macro_rules! impl_spim_ext {
+        ram_slice_check(words)?;
 
-            self.do_spi_dma_transfer(data_ptr as u32,data_len as u32,data_ptr as u32,data_len as u32,|_|{})?;
+        words.chunks(EASY_DMA_SIZE).try_for_each(|chunk| {
+            self.do_spi_dma_transfer(
+                DmaSlice::from_slice(chunk),
+                DmaSlice::from_slice(chunk),
+                |_| {},
+            )
+        })?;
 
-        }
         Ok(words)
     }
 }
-impl<T> embedded_hal::blocking::spi::Write<u8> for Spim<T> where T: SpimExt
+
+impl<T> embedded_hal::blocking::spi::Write<u8> for Spim<T>
+where
+    T: SpimExt,
 {
-   type Error = Error;
+    type Error = Error;
 
     fn write<'w>(&mut self, words: &'w [u8]) -> Result<(), Error> {
-        let p = words.as_ptr() as usize;
         // Mask on segment where Data RAM is located on nrf52840 and nrf52832
         // Upper limit is choosen to entire area where DataRam can be placed
-        if SRAM_LOWER <= p && p < SRAM_UPPER {
-            let mut offset:usize = 0;
-            while offset < words.len() {
-                let data_len = min(EASY_DMA_SIZE, words.len() - offset);
-                let data_ptr = offset + (words.as_ptr() as usize);
-                offset += data_len;
-
-                // setup spi dma tx buffer and 0 for read buffer length
-                self.do_spi_dma_transfer(data_ptr as u32,data_len as u32,0,0,|_|{})?;
-            }
+        let needs_copy = !slice_in_ram(words);
+
+        let chunk_sz = if needs_copy {
+            FORCE_COPY_BUFFER_SIZE
         } else {
-            // Force copy from flash mode.
-            let blocksize = min(EASY_DMA_SIZE,FORCE_COPY_BUFFER_SIZE);
-            let mut buffer:[u8;FORCE_COPY_BUFFER_SIZE] = [0;FORCE_COPY_BUFFER_SIZE];
-            let mut offset:usize = 0;
-            while offset < words.len() {
-                let data_len = min(blocksize, words.len() - offset);
-                for i in 0..data_len{
-                    buffer[i] = words[offset+i];
-                }
-                offset += data_len;
-                // setup spi dma tx buffer and 0 for read buffer length
-                self.do_spi_dma_transfer(buffer.as_ptr() as u32,data_len as u32,0,0,|_|{})?;
-            }
-        }
-        Ok(())
+            EASY_DMA_SIZE
+        };
+
+        let step = if needs_copy {
+            Self::spi_dma_copy
+        } else {
+            Self::spi_dma_no_copy
+        };
 
+        words.chunks(chunk_sz).try_for_each(|c| step(self, c))
     }
 }
-impl<T> Spim<T> where T: SpimExt {
+impl<T> Spim<T>
+where
+    T: SpimExt,
+{
+    fn spi_dma_no_copy(&mut self, chunk: &[u8]) -> Result<(), Error> {
+        self.do_spi_dma_transfer(DmaSlice::from_slice(chunk), DmaSlice::null(), |_| {})
+    }
+
+    fn spi_dma_copy(&mut self, chunk: &[u8]) -> Result<(), Error> {
+        let mut buf = [0u8; FORCE_COPY_BUFFER_SIZE];
+        buf[..chunk.len()].copy_from_slice(chunk);
```

Comment:
TODO: make sure `chunk.len() <= FORCE_COPY_BUFFER_SIZE`, or at least document that limitation

Reply:
Interesting concern, `copy_from_slice` will panic it length are not equal, is it possible to compile misconfigured?